Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -266,114 +239,187 @@ final class ConnectionPool {
 
-        /// The pool this provider belongs to
-        private let parentPool: ConnectionPool
+        private var active: Bool = true
+
+        /// Opened connections that are available
+        var availableConnections: CircularBuffer<Connection> = .init(initialCapacity: 8)
+
+        /// Consumers that weren't able to get a new connection without exceeding
+        /// `maximumConcurrentConnections` get a `Future<Connection>`
+        /// whose associated promise is stored in `Waiter`. The promise is completed
+        /// as soon as possible by the provider, in FIFO order.
+        var waiters: CircularBuffer<Waiter> = .init(initialCapacity: 8)
+
+        /// Number of opened or opening connections, used to keep track of all connections and enforcing `maximumConcurrentConnections` limit.
+        var openedConnectionsCount: Int = 0
 
         /// Creates a new `HTTP1ConnectionProvider`
         ///
         /// - parameters:
         ///     - key: The `Key` (host, scheme, port) this provider is associated to
         ///     - configuration: The client configuration used globally by all requests
         ///     - initialConnection: The initial connection the pool initializes this provider with
-        ///     - parentPool: The pool this provider belongs to
-        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
+        ///     - pool: The pool this provider belongs to
+        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, pool: ConnectionPool) {
             self.eventLoop = eventLoop
             self.configuration = configuration
             self.key = key
-            self.parentPool = parentPool
-            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+            self.pool = pool
         }
 
         deinit {
-            assert(self.state.activity == .closed, "Non closed on deinit")
-            assert(self.state.availableConnections.isEmpty, "Available connections should be empty before deinit")
-            assert(self.state.leased == 0, "All leased connections should have been returned before deinit")
-            assert(self.state.waiters.count == 0, "Waiters on deinit: \(self.state.waiters)")
+            assert(self.waiters.isEmpty)
+            assert(self.availableConnections.isEmpty)
+            assert(self.openedConnectionsCount == 0)
         }
 
         var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
+            return "HTTP1ConnectionProvider { key: \(self.key) }"
         }
 
-        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
+        var isActive: Bool {
```

Comment:
I think this is confusing:

- `self.active` when already locked
- `self.isActive` when currently not locked?

Also, as mentioned above: what does "active" mean in this context?